added initial .travis.yml

krichter722 opened the pull request:
This should help reveal issues like https://travis-ci.org/krichter722/tika/jobs/258660634 which are apparently not covered by the current CI.

I experience the test failure locally on Ubuntu 17.04 as well. Imo it's better to merge the Travis CI script first which allows you to fix the build and test issues and then I'll report them if they still occur locally in a while.

grossws commented:
[Jenkins at builds.apache.org](https://builds.apache.org/view/T/view/Tika/job/Tika-trunk/1338/org.apache.tika$tika-parsers/console) catched that and it was fixed in a51add2afe475ba8d3fb5847c97f52b1eb71deea by @Gagravarr.

Personally I don't see a reason to add and maintain yet another CI configuration if current works reasonably. I'm closing this for now since "are apparently not covered by the current CI" statement is technically false.

krichter722 commented:
> Jenkins at builds.apache.org catched that and it was fixed in a51add2 by @Gagravarr.

> Personally I don't see a reason to add and maintain yet another CI configuration if current works reasonably. I'm closing this for now since "are apparently not covered by the current CI" statement is technically false.

OK, I agree. However it's strange that failing builds end up in `master`. I recommend to change the practice and not push into master unless the build is stable (with fast-forward merges and rebasing any pollution of the git history can be avoided). Today I encountered

<details>
  <summary>Build failure in OOXMLParserTest.testBigIntegersWGeneralFormat</summary>
<pre><![CDATA[
  Failed tests: 
  OOXMLParserTest.testBigIntegersWGeneralFormat:1350->TikaTest.assertContains:102 1.23456789012345E+15</td>	<td>1.23456789012345E+15 not found in:
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta name="date" content="2016-07-22T11:32:25Z" />
<meta name="extended-properties:AppVersion" content="16.0300" />
<meta name="dc:creator" content="Allison, Timothy B." />
<meta name="extended-properties:Company" content="" />
<meta name="dcterms:created" content="2016-06-29T16:29:27Z" />
<meta name="Last-Modified" content="2016-07-22T11:32:25Z" />
<meta name="dcterms:modified" content="2016-07-22T11:32:25Z" />
<meta name="Last-Save-Date" content="2016-07-22T11:32:25Z" />
<meta name="protected" content="false" />
<meta name="meta:save-date" content="2016-07-22T11:32:25Z" />
<meta name="Application-Name" content="Microsoft Excel" />
<meta name="modified" content="2016-07-22T11:32:25Z" />
<meta name="Content-Type" content="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" />
<meta name="X-Parsed-By" content="org.apache.tika.parser.DefaultParser" />
<meta name="X-Parsed-By" content="org.apache.tika.parser.microsoft.ooxml.OOXMLParser" />
<meta name="creator" content="Allison, Timothy B." />
<meta name="meta:author" content="Allison, Timothy B." />
<meta name="meta:creation-date" content="2016-06-29T16:29:27Z" />
<meta name="extended-properties:Application" content="Microsoft Excel" />
<meta name="meta:last-author" content="Allison, Timothy B." />
<meta name="Creation-Date" content="2016-06-29T16:29:27Z" />
<meta name="Last-Author" content="Allison, Timothy B." />
<meta name="Application-Version" content="16.0300" />
<meta name="Author" content="Allison, Timothy B." />
<meta name="publisher" content="" />
<meta name="dc:publisher" content="" />
<title></title>
</head>
<body><div><h1>Sheet1</h1>
<table><tbody><tr>	<td>123456789012345</td>	<td>123456789012346</td></tr>
<tr>	<td>1,23456789012345E+15</td>	<td>1,23456789012345E+15</td></tr>
<tr />
</tbody></table>
</div>
</body></html>
  OOXMLParserTest.testXLSBVarious:1552->TikaTest.assertContains:102 <td>13.1211231321</td> not found in:
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta name="date" content="2017-03-10T14:58:49Z" />
<meta name="extended-properties:AppVersion" content="16.0300" />
<meta name="dc:creator" content="Allison, Timothy B." />
<meta name="extended-properties:Company" content="" />
<meta name="dcterms:created" content="2017-03-09T12:24:26Z" />
<meta name="Last-Modified" content="2017-03-10T14:58:49Z" />
<meta name="dcterms:modified" content="2017-03-10T14:58:49Z" />
<meta name="Last-Save-Date" content="2017-03-10T14:58:49Z" />
<meta name="protected" content="false" />
<meta name="meta:save-date" content="2017-03-10T14:58:49Z" />
<meta name="Application-Name" content="Microsoft Excel" />
<meta name="modified" content="2017-03-10T14:58:49Z" />
<meta name="Content-Type" content="application/vnd.ms-excel.sheet.binary.macroenabled.12" />
<meta name="X-Parsed-By" content="org.apache.tika.parser.DefaultParser" />
<meta name="X-Parsed-By" content="org.apache.tika.parser.microsoft.ooxml.OOXMLParser" />
<meta name="creator" content="Allison, Timothy B." />
<meta name="meta:author" content="Allison, Timothy B." />
<meta name="meta:creation-date" content="2017-03-09T12:24:26Z" />
<meta name="extended-properties:Application" content="Microsoft Excel" />
<meta name="meta:last-author" content="Allison, Timothy B." />
<meta name="Creation-Date" content="2017-03-09T12:24:26Z" />
<meta name="Last-Author" content="Allison, Timothy B." />
<meta name="X-TIKA:origResourceName" content="C:\Users\tallison\Desktop\working\xlsb\" />
<meta name="Application-Version" content="16.0300" />
<meta name="Author" content="Allison, Timothy B." />
<meta name="publisher" content="" />
<meta name="dc:publisher" content="" />
<title></title>
</head>
<body><div><h1>mySheet1</h1>
<table><tbody><tr>	<td>String</td>	<td>This is a string</td></tr>
<tr>	<td>integer</td>	<td>13</td></tr>
<tr>	<td>float</td>	<td>13,1211231321</td></tr>
<tr>	<td>currency</td>	<td>$   3.03</td></tr>
<tr>	<td>percent</td>	<td>20%</td></tr>
<tr>	<td>float 2</td>	<td>13,12</td></tr>
<tr>	<td>long int</td>	<td>123456789012345</td></tr>
<tr>	<td>longer int</td>	<td>1,23456789012345E+15</td>	<td><br />
Allison, Timothy B.: Allison, Timothy B.:
test comment2
</td></tr>
<tr>	<td>fraction</td>	<td>1/4</td></tr>
<tr>	<td>date</td>	<td>3/9/17</td></tr>
<tr>	<td>comment</td>	<td>contents<br />
Allison, Timothy B.: Allison, Timothy B.:
test comment
</td></tr>
<tr>	<td>hyperlink</td>	<td>tika_link</td></tr>
<tr>	<td>formula</td>	<td>4</td>	<td>2</td></tr>
<tr>	<td>formulaErr</td>	<td>ERROR</td></tr>
<tr>	<td>formulaFloat</td>	<td>0,5</td>	<td>March</td>	<td>April</td></tr>
<tr>	<td>customFormat1</td>	<td>   46/1963</td>	<td>merchant1</td>	<td>1</td>	<td>3</td></tr>
<tr>	<td>customFormat2</td>	<td>  3/128</td>	<td>merchant2</td>	<td>2</td>	<td>4</td></tr>
<tr>	<td>text test</td></tr>
<tr>	<td><br />
Allison, Timothy B.: Allison, Timothy B.:
comment1
</td></tr>
<tr>	<td><br />
Allison, Timothy B.: Allison, Timothy B.:
comment2
</td></tr>
<tr>	<td><br />
Allison, Timothy B.: Allison, Timothy B.:
comment3
</td></tr>
<tr>	<td>the</td>	<td><br />
Allison, Timothy B.: Allison, Timothy B.:
comment4 (end of row)
</td></tr>
<tr>	<td>the</td>	<td><br />
Allison, Timothy B.: Allison, Timothy B.:
comment5 between cells
</td>	<td>quick</td></tr>
<tr>	<td>comment6<br />
Allison, Timothy B.: Allison, Timothy B.:
comment6 actually in cell
</td></tr>
<tr>	<td><br />
Allison, Timothy B.: Allison, Timothy B.:
comment7 end of file
</td></tr>
<tr>	<td><br />
Allison, Timothy B.: Allison, Timothy B.:
comment8 end of file</td></tr>
</tbody></table>
<p>OddLeftHeader OddCenterHeader OddRightHeader</p>
<p>EvenLeftHeader EvenCenterHeader EvenRightHeader
</p>
<p>FirstPageLeftHeader FirstPageCenterHeader FirstPageRightHeader</p>
<p>OddLeftFooter OddCenterFooter OddRightFooter</p>
<p>EvenLeftFooter EvenCenterFooter EvenRightFooter</p>
<p>FirstPageLeftFooter FirstPageCenterFooter FirstPageRightFooter</p>
<p>test textbox
</p>
<a href="http://lucene.apache.org/">http://lucene.apache.org/</a><p>myChartTitle</p>
<p />
merchant1	March	April	1	3	merchant2	March	April	2	4	<p />
<p />
<p />
<p />
<p>test WordArt</p>
<p>myChartTitle</p>
<p />
merchant1	March	April	1	3	merchant2	March	April	2	4	<p />
<p />
<p />
<p />
<p>myChartTitle</p>
<p />
merchant1	March	April	1	3	merchant2	March	April	2	4	<p />
<p />
<p />
<p />
<a href="http://tika.apache.org/">http://tika.apache.org/</a></div>
<div class="package-entry" /><div class="package-entry" /><div class="package-entry" /></body></html>
]]>
</pre>
</details>

(afaik github.com doesn't parse that right) which makes it hard to start commiting to the project because one needs a building project in order to start working because suggesting improvements of the base of the latest release produces unnecessary merge work.

grossws commented:
> However it's strange that failing builds end up in master. I recommend to change the practice and not push or merge into master unless the build is stable (with fast-forward merges and rebasing any pollution of the git history can be avoided)

`master` branch in out case is development branch, not a release one. It has both historical and organizational reasons: we migrated from `svn` where branches and tags are quite expensive, so we haven't adopted gitflow or something similar which use motto `master is stable and always builds` which is relevant for software with CD-pipeline in the first place. Organizational part is even harder, we are group of individuals from different companies and different places around the planet (same applies to all Apache Software Foundation projects). It leads to asynchronous communication with large latencies and more complicated repo guidelines would get a lot of valuable time: both PMC/committers' and contributors'. So other global or organizational tasks have higher prio to me. I think github is not the best place for such discussions, so if you want to continue it and bring to attention of other Tika devs I'll suggest to mail to dev@tika.apache.org (see docs at http://tika.apache.org/mail-lists.html to subscribe).

> which makes it hard to start commiting to the project because one needs a building project in order to start working because suggesting improvements of the base of the latest release produces unnecessary merge work.

Usually it isn't that bad) Also, you can base your patch on release tag, test it and rebase onto `master` before merge. If a build from release tag fails it usually means some exotic environment which slipped off out attention and probably to be fixed in `master` to avoid such issues in next release.

krichter722 commented:
> Organizational part is even harder, we are group of individuals from different companies and different places around the planet (same applies to all Apache Software Foundation projects).

That applies to almost any FLOSS project that exists and is hardly worth mentioning. Local distribution and communication latency isn't a good reason to allow failing builds on `master`. The discussion could continue on the mailing list, but I see no sense in it since having a main development branch that doesn't build reliably is a no-go for contributions from my side and it's basically a yes-no-decision - I didn't want to contribute much, but other devs might and it'd be a pity to loose their contributions - it's your decision and I understand you priorities.

> Also, you can base your patch on release tag, test it and rebase onto master before merge.

That causes unnecessary work and should be discouraged. It's a valid workaround for the current situation.

tballison commented:
> but I see no sense in it since having a main development branch that doesn't build reliably is a no-go for contributions from my side

I completely agree.  Trunk should build...end of story.  This is in fact one of the reasons that I haven't been eager to contribute to Solr.

We found the bug you mentioned during the last release process.  This was a locale bug that was my fault, and which I didn't see on my local machine and which our CI process didn't uncover.  That we didn't catch this earlier is symptomatic of the few number of devs actively building snapshots, so yes, we absolutely need more!

If trunk doesn't build, please do let us know on dev@ or on JIRA.  I share your concern 100% that this is a big problem and will discourage participation if we don't fix it.

grossws commented:
@

grossws commented:
@krichter722, I totally agree with you and @tballison that `master` should be stable enough to avoid wasting contributors' time to get it building in their environment. Unfortunately it isn't so yet and should be fixed. My thoughts about making patch against release tag is mere workaround. 

All I said above isn't Tika PMC position but my personal view on what I can do and plan to do in short term future.

> BTW, do you want to have a report for the test failure mentioned above?

Of course, if you would. We appreciate bug reports. I couldn't reproduce it on my laptop with current `master` (4455a6f087bbd52767e55478e9d51fdb39adbce6)
